sdcm/cluster_gce.py

```diff
@@ -341,42 +341,59 @@ def _create_instance(self, node_index, dc_idx, spot=False):
-                    (failed_instance := self._get_instances_by_name(dc_idx=dc_idx, name=name)):
-                try:
-                    self._gce_services[dc_idx].destroy_node(node=failed_instance, destroy_boot_disk=True)
-                # if the instance is destroyed already - ignore the exception
-                # otherwise - raise exception
-                except ResourceNotFoundError as exc:
-                    if 'notFound' not in str(exc):
-                        raise
 
-            create_node_params['ex_preemptible'] = spot = False
-            instance = self._gce_services[dc_idx].create_node(**create_node_params)
+        instance = self._create_node_with_retries(name=name,
+                                                  dc_idx=dc_idx,
+                                                  create_node_params=create_node_params,
+                                                  spot=spot)
 
         self.log.info('Created %s instance %s', 'spot' if spot else 'on-demand', instance)
         if gce_job_default_timeout:
             self.log.info('Restore default job timeout %s' % gce_job_default_timeout)
             self._gce_services[dc_idx].connection.timeout = gce_job_default_timeout
         return instance
 
+    @retrying(n=3,
+              sleep_time=900,
+              allowed_exceptions=(GoogleBaseError, ResourceNotFoundError),
+              message="Retrying to create a GCE node...",
+              raise_on_exceeded=True)
+    def _create_node_with_retries(self,
+                                  name: str,
+                                  dc_idx: int,
+                                  create_node_params: dict[str, Any],
+                                  spot: bool = False) -> GCENode:
+        try:
+            instance = self._gce_services[dc_idx].create_node(**create_node_params)
+            return instance
+        except GoogleBaseError as gbe:
+            if not spot:
+                raise gbe
+
+            #  attempt to destroy if node did not start due to preemption
+            if "Instance failed to start due to preemption" in str(gbe):
+                try:
+                    self._destroy_node(name=name, dc_idx=dc_idx)
+                except ResourceNotFoundError:
+                    LOGGER.warning("Attempted to destroy node: {name: %s, idx:%s}, but could not find it. "
+                                   "Possibly the node was destroyed already.", name, str(dc_idx))
+
+            create_node_params['ex_preemptible'] = spot = False
+
+            raise gbe
+
     def _create_instances(self, count, dc_idx=0):
         spot = 'spot' in self.instance_provision
         instances = []
         for node_index in range(self._node_index + 1, self._node_index + count + 1):
             instances.append(self._create_instance(node_index, dc_idx, spot))
         return instances
 
+    def _destroy_node(self, name: str, dc_idx: int) -> bool:
```

Comment:
Do you think `_destroy_instance` will be better name for this method?

Reply:
Yes, that's better. Fixed.